### PR TITLE
[examples] add NXP JN5189 platform

### PR DESCRIPTION
**Pull request by doru91**
Pull-request for NXP JN5189 platform.

### CODE REVIEW

File: examples/Makefile-jn5189 [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+CONFIG_FILE      = OPENTHREAD_PROJECT_CORE_CONFIG_FILE='\"openthread-core-jn5189-config.h\"'
+CONFIG_FILE_PATH = $(AbsTopSourceDir)/examples/platforms/jn5189/
+SIGN_IMAGE_PATH = $(AbsTopSourceDir)/third_party/nxp/JN5189/ImageSigning/
+
+COMMONCFLAGS                          := \
+    -fdata-sections                      \
+    -ffunction-sections                  \
+    -Os                                  \
+    -g                                   \
+	-DCPU_JN518X                         \
```
Comment (jwhui) [inline examples/Makefile-jn5189]: nit: use spaces instead of tabs

File: examples/platforms/Makefile.am [hunk 2]
```diff
@@ -39,6 +39,7 @@ DIST_SUBDIRS                            = \
     efr32mg13                             \
     efr32mg21                             \
     gp712                                 \
+	jn5189                                \
```
Comment (jwhui) [inline examples/platforms/Makefile.am]: nit: use spaces instead of tabs

File: examples/platforms/jn5189/Makefile.am [hunk 3]
```diff
@@ -0,0 +1,129 @@
+#
+#  Copyright (c) 2019, The OpenThread Authors.
+#  Copyright (c) 2019, NXP.
```
Comment (jwhui) [inline examples/platforms/jn5189/Makefile.am]: The only copyright holder should be "The OpenThread Authors"

File: examples/platforms/jn5189/Makefile.am [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+#  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+#  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+#  POSSIBILITY OF SUCH DAMAGE.
+#
+
+include $(abs_top_nlbuild_autotools_dir)/automake/pre.am
+
+lib_LIBRARIES                                                                                      = \
+    libopenthread-jn5189_plat.a                                                                      \
+	libopenthread-jn5189_sdk.a
```
Comment (jwhui) [inline examples/platforms/jn5189/Makefile.am]: nit: use spaces instead of tabs

File: examples/platforms/jn5189/Makefile.platform.am [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+#
+
+#
+# JN5189 platform-specific Makefile
+#
+
+LDADD_COMMON                                                               += \
+    $(top_builddir)/examples/platforms/jn5189/libopenthread-jn5189_plat.a     \
+    $(top_builddir)/examples/platforms/jn5189/libopenthread-jn5189_sdk.a      \
+	$(top_srcdir)/third_party/nxp/JN5189/libraries/libMiniMac.a               \
```
Comment (jwhui) [inline examples/platforms/jn5189/Makefile.platform.am]: nit: use spaces instead of tabs

File: examples/platforms/jn5189/Makefile.platform.am [hunk 6]
```diff
@@ -0,0 +1,44 @@
+#
+#  Copyright (c) 2019, The OpenThread Authors.
+#  Copyright (c) 2019, NXP.
```
Comment (jwhui) [inline examples/platforms/jn5189/Makefile.platform.am]: The only copyright holder should be "The OpenThread Authors"

File: examples/platforms/jn5189/Makefile.am [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+    flash.c                                                                                          \
+    logging.c                                                                                        \
+    misc.c                                                                                           \
+    radio.c                                                                                          \
+    entropy.c                                                                                        \
+    system.c                                                                                         \
+    uart.c                                                                                           \
+    settings_jn5189.c                                                                                \
+    @top_builddir@/third_party/nxp/JN5189/drivers/fsl_assert.c                                       \
+	@top_builddir@/third_party/nxp/JN5189/crypto/aes_alt.c                                           \
```
Comment (jwhui) [inline examples/platforms/jn5189/Makefile.am]: nit: use spaces instead of tabs

File: examples/platforms/jn5189/Makefile.am [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+    @top_builddir@/third_party/nxp/JN5189/crypto/ksdk_mbedtls.c                                      \
+    $(NULL)
+
+libopenthread_jn5189_sdk_a_SOURCES                                                                 = \
+    @top_builddir@/third_party/nxp/JN5189/startup_JN5189.c                                           \
+    @top_builddir@/third_party/nxp/JN5189/system_JN5189.c                                            \
+    @top_builddir@/third_party/nxp/JN5189/dk6_jn5180/pin_mux.c                                       \
+    @top_builddir@/third_party/nxp/JN5189/dk6_jn5180/clock_config.c                                  \
+    @top_builddir@/third_party/nxp/JN5189/drivers/fsl_gpio.c                                        \
```
Comment (jwhui) [inline examples/platforms/jn5189/Makefile.am]: nit: align '\'

File: examples/platforms/jn5189/README.md [hunk 9]
```diff
@@ -0,0 +1,138 @@
+# OpenThread on NXP JN5189 Example
+
+This directory contains example platform drivers for the [NXP JN5189][jn5189]
```
Comment (jwhui) [inline examples/platforms/jn5189/README.md]: missing references for both jn5189 and jn5189-dk006

File: examples/platforms/jn5189/Makefile.am [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+    @top_builddir@/third_party/nxp/JN5189/drivers/aes/fsl_aes.c                                      \
+    @top_builddir@/third_party/nxp/JN5189/drivers/sha/fsl_sha.c                                      \
+    $(NULL)
+
+libopenthread_jn5189_plat_a_SOURCES                                                                = \
+    $(PLATFORM_SOURCES)                                                                              \
+    $(NULL)
+
+PRETTY_FILES                                                                                       = \
+    $(PLATFORM_SOURCES)                                                                              \
```
Comment (jwhui) [inline examples/platforms/jn5189/Makefile.am]: There are some `third_party/nxp` files included in `PLATFORM_SOURCES`. Do you really want OpenThread's make pretty running on `third_party/nxp` files?

File: examples/platforms/jn5189/openthread-core-jn5189-config.h [hunk 11]
```diff
@@ -0,0 +1,195 @@
+/*
+ *  Copyright (c) 2019, The OpenThread Authors.
+ *  Copyright (c) 2019, NXP.
```
Comment (jwhui) [inline examples/platforms/jn5189/openthread-core-jn5189-config.h]: The only copyright holder should be "The OpenThread Authors"

**Comment by jwhui** [comment]
@doru91, thank you for contributing this new platform!

Can we add this platform to the Travis build scripts in [.travis/script.sh](https://github.com/openthread/openthread/blob/master/.travis/script.sh)?

File: src/core/common/debug.hpp [hunk 12]
```diff
@@ -77,6 +77,7 @@
 #undef assert
 #endif
 
+#ifndef assert
```
Comment (jwhui) [inline src/core/common/debug.hpp]: Is this necessary? Just a few lines above, there is a check and `#undef assert`.

File: examples/Makefile-jn5189 (hunk 1 above)
Comment (doru91) [reply examples/Makefile-jn5189]: Fixed.

File: examples/platforms/Makefile.am (hunk 2 above)
Comment (doru91) [reply examples/platforms/Makefile.am]: Fixed.

File: examples/platforms/jn5189/Makefile.am (hunk 3 above)
Comment (doru91) [reply examples/platforms/jn5189/Makefile.am]: Fixed.

File: examples/platforms/jn5189/Makefile.am (hunk 4 above)
Comment (doru91) [reply examples/platforms/jn5189/Makefile.am]: Fixed.

File: examples/platforms/jn5189/Makefile.platform.am (hunk 5 above)
Comment (doru91) [reply examples/platforms/jn5189/Makefile.platform.am]: Fixed.

File: examples/platforms/jn5189/Makefile.platform.am (hunk 6 above)
Comment (doru91) [reply examples/platforms/jn5189/Makefile.platform.am]: Fixed.

File: examples/platforms/jn5189/Makefile.am (hunk 7 above)
Comment (doru91) [reply examples/platforms/jn5189/Makefile.am]: Fixed.

File: examples/platforms/jn5189/Makefile.am (hunk 8 above)
Comment (doru91) [reply examples/platforms/jn5189/Makefile.am]: Fixed.

File: examples/platforms/jn5189/README.md (hunk 9 above)
Comment (doru91) [reply examples/platforms/jn5189/README.md]: Added references.

File: examples/platforms/jn5189/Makefile.am (hunk 10 above)
Comment (doru91) [reply examples/platforms/jn5189/Makefile.am]: No, I reverted the files from third_party/nxp.

File: examples/platforms/jn5189/openthread-core-jn5189-config.h (hunk 11 above)
Comment (doru91) [reply examples/platforms/jn5189/openthread-core-jn5189-config.h]: Fixed.

**Comment by doru91** [comment]
> @doru91, thank you for contributing this new platform!
> 
> Can we add this platform to the Travis build scripts in [.travis/script.sh](https://github.com/openthread/openthread/blob/master/.travis/script.sh)?

Sure.

File: src/core/common/debug.hpp (hunk 12 above)
Comment (doru91) [reply src/core/common/debug.hpp]: It is not necessary, I removed it.

File: examples/platforms/jn5189/Makefile.am (hunk 10 above)
Comment (doru91) [reply examples/platforms/jn5189/Makefile.am]: The problem now is that the Travis CI build is failing due to pretty check errors caused by third_party/nxp files. How could I fix that?

Comment (jwhui) [reply examples/platforms/jn5189/Makefile.am]: #4660 should help address this.

Comment (jwhui) [reply examples/platforms/jn5189/Makefile.am]: @doru91, I think you'll need to run `./script/make-pretty clang` again to resolve the code format issues.

Comment (doru91) [reply examples/platforms/jn5189/Makefile.am]: Thanks, @jwhui. I run again the make-pretty script and now the Travis build is fixed.

Comment (doru91) [reply examples/platforms/jn5189/Makefile.am]: @jwhui, is there any other change needed in order to have this pull request approved?

Comment (jwhui) [reply examples/platforms/jn5189/Makefile.am]: Had to resolve a couple conflicts in `configure.ac`. This should be ready once the build checks pass.

**Pull request merged**